python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+                    dwn_tester.setStatus('skipped dependency', dwn_tester.bucket_skip)
+
+                    # Add this deleted tester to a set
+                    deleted_tester_set.add(dwn_tester)
+
+                    # Ask a status thread to print our results
+                    self.statusGo(downstream_test_name)
+
+                # Like Note 1 above, the same applies here. We need to _attempt_ to delete the node, but not
+                # tell the harness multiple times that this test was skipped. This 'downsteam' node just
+                # happens to exist in mulitple places in the graph. Exp:
+                # A     B
+                #  \   /
+                #    C  <--- Downstream 'C' node mentioned in both A and B.
+                if not dwn_tester.skipPrereqs(self.options):
+                    dag_object.delete_node_if_exists(downstream_test_name)
+
+        # Return the set of tester objects we deleted
+        return deleted_tester_set
+
+    # Loop through the testers and add them to the runner pool for execution
+    def schedule(self, testers):
+        # We didn't recieve anything... pesky TestHarness
+        if testers == []:
+            return
+
+        # Immediately test for race conditions and return if condition exists
+        if self.checkRaceCondition(testers):
+            return
+
+        # Local storage of skipped tester objects
+        skipped_or_failed_tests = set([])
+
+        # Local storage of downstream deleted tester objects
+        downstream_deleted = set([])
+
+        # Local DAG
+        tester_dag = dag.DAG()
+
+        # Create the DAG, skipped_testers, and just about anything we can during this iteration loop
+        for tester in testers:
+            with self.thread_lock:
+                # Create tester name to object map
+                self.scheduled_name_to_object[tester.getTestName()] = tester
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
+                tester_dag.add_node(tester.getTestName())
+
+            # Loop through prereqs and add leaves/nodes where necessary
+            for prereq in tester.getPrereqs():
+
+                # Discover unknown prereqs
+                if prereq not in [x.getTestName() for x in testers]:
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    skipped_or_failed_tests.add(tester)
+                    continue
+
+                if not tester_dag.node_exists(prereq):
+                    # Create a node for this tester
+                    tester_dag.add_node(prereq)
+
+                # Once you add an edge to a node, the DAG arranges the nodes accordingly
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors (tester status has already be applied)
```

Comment:
Where do we find cyclic tests? Is this in the checkRaceCondition stuff? Maybe we should just combine that into building the DAG.